support-workers/src/main/scala/com/gu/support/workers/lambdas/CreateZuoraSubscription.scala

```diff
@@ -323,23 +323,42 @@ object DigitalSubscriptionGiftRedemption {
     val codeValidator = new GiftCodeValidator(zuoraService)
     for {
       codeValidation <- codeValidator.getStatus(redemptionData.redemptionCode, Some(state.requestId.toString))
-      subIdUpdateAction <- codeValidation match {
-        case ValidGiftCode(subscriptionId) => Future.successful((
+      termDates <- codeValidation match {
+        case ValidGiftCode(subscriptionId) => doZuoraUpdates(
           subscriptionId,
-          zuoraService.updateSubscriptionRedemptionData(subscriptionId, state.requestId.toString, state.user.id, LocalDate.now(), _)
-        ))
-        case CodeRedeemedInThisRequest(subscriptionId) => Future.successful(subscriptionId, (_: Int) => Future.successful(UpdateRedemptionDataResponse(true)))
+          zuoraService.updateSubscriptionRedemptionData(subscriptionId, state.requestId.toString, state.user.id, LocalDate.now(), _),
+          zuoraService.setupRevenueRecognition,
+          zuoraService,
+          catalogService
+        )
+        case CodeRedeemedInThisRequest(subscriptionId) => doZuoraUpdates(
+          subscriptionId,
+          (_: Int) => Future.successful(ZuoraSuccessOrFailureResponse(success = true, None)),
+          (_: Subscription, _:TermDates) => Future.successful(()),
```

Comment:
I'm not sure this really is idempodent now.  The lambda can fail at any point on the first iteration - e.g. zuora could time out when doing the upddate identity id call, yet succeed.  Then when the lambda runs again it will be redemmed in this request, but the revenue distribution would still need to happen.
We would need to redo the distribution really, to be sure, or check and do if necessary.

Reply:
I've updated this so that all code paths now call `setupRevenueRecognition` and that function checks whether the distribute revenue call has already happened before calling it again